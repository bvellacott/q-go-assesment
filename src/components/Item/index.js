const Item = ({ content, complete, remove, toggleComplete }) => (
	<span className={`item ${complete ? 'item_complete' : 'item_incomplete'}`}>
		{toggleComplete && (
		<button className="item_toggleCompleteButton" onClick={toggleComplete}>
			{complete ? 'set to done' : 'set to incomplete'}
		</button>)}
		<span className="item_content">{content}</span>
		{remove && (
		<button className="item_removeButton" onClick={remove}>X</button>)}
	</span>)

export default Item
