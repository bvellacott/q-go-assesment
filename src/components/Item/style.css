.item_content {
	margin-right: 1rem;
}

.item_complete .item_content {
	color: #5f5;
}

.item_incomplete .item_content {
	color: #f85;
}