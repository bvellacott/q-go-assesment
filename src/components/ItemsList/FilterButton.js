const FilterButton = () => { throw new Error('not implemented') }
