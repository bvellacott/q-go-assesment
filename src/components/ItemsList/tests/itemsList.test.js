import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import FilterButton from '../FilterButton';
import ItemsListContainer, { ItemsList } from '../index';
import Item from '../../Item'
import configureStore from 'redux-mock-store'

let mockSetFilter
let mockToggleTodoComplete
let mockRemoveTodo
jest.mock('../../../logic/actions', () => ({
  setFilter: (...args) => mockSetFilter(...args),
  toggleTodoComplete: (...args) => mockToggleTodoComplete(...args),
  removeTodo: (...args) => mockRemoveTodo(...args),
}), { virtual: true })

const mockStore = configureStore([])

const defaultProps = {
  items: [],
  filterAll: () => {},
  filterComplete: () => {},
  filterIncomplete: () => {},
  toggleTodoComplete: () => {},
  removeTodo: () => {},
};

describe('ItemsList', () => {
  it('renders without crashing', () => {
    shallow(<ItemsList {...defaultProps} />);
  });

  it('should display warning message if no items', () => {
    const renderedItem = shallow(<ItemsList {...defaultProps} items={[]} />);
    expect(renderedItem.find('#items-missing')).toHaveLength(1);
  });

  it('should not display warning message if items are present', () => {
    const items = [{ id: 1, content: 'Test 1' }];
    const renderedItem = shallow(<ItemsList {...defaultProps} items={items} />);
    expect(renderedItem.find('#items-missing')).toHaveLength(0);
  });

  it('should render items as list items', () => {
    const items = [{ id: 1, content: 'Test 1' }, { id: 2, content: 'Test 2' }];
    const renderedItem = shallow(<ItemsList {...defaultProps} items={items} />);
    expect(renderedItem.find('.itemsList-ul li')).toHaveLength(2);
  });

  it('Renders item component and passes toggle and remove actions', () => {
    const toggleTodoComplete = jest.fn()
    const removeTodo = jest.fn()

    const items = [{ id: 1, content: 'Test 1' }]
    const renderedItem = shallow(<ItemsList 
      {...defaultProps} 
      items={items}
      toggleTodoComplete={toggleTodoComplete}
      removeTodo={removeTodo}
    />)
    const item = renderedItem.find(Item)
    expect(item).toHaveLength(1)

    let itemProps = item.props()
    expect(typeof itemProps.toggleComplete).toEqual('function')
    expect(typeof itemProps.remove).toEqual('function')

    itemProps.toggleComplete()
    expect(toggleTodoComplete).toHaveBeenCalledWith(1)

    itemProps.remove()
    expect(removeTodo).toHaveBeenCalledWith(1)
  });

  it('renders filter button and correct handlers', () => {
    const wrapper = shallow(<ItemsList 
      {...defaultProps} 
      filterAll={() => 'all'}
      filterComplete={() => 'complete'}
      filterIncomplete={() => 'incomplete'}
    />);

    const filterButtons = wrapper.find(FilterButton)
    expect(filterButtons.length).toEqual(3)
    expect(filterButtons.at(0).props().onClick()).toEqual('all')
    expect(filterButtons.at(1).props().onClick()).toEqual('complete')
    expect(filterButtons.at(2).props().onClick()).toEqual('incomplete')

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('passes items and merged filter actions from store', () => {
    const store = mockStore({
      todos: {
        items: [1, 2, 3],
      }
    })
    mockSetFilter = filter => ({ type: 'test_set_filter', filter })
    mockToggleTodoComplete = id => ({ type: 'test_toggle_complete', id })
    mockRemoveTodo = id => ({ type: 'test_remove_todo', id })

    const wrapper = shallow(<ItemsListContainer store={store} />);

    const passedProps = wrapper.props()
    expect(passedProps.items).toEqual([1, 2, 3])

    expect(typeof passedProps.filterAll).toEqual('function')
    expect(typeof passedProps.filterComplete).toEqual('function')
    expect(typeof passedProps.filterIncomplete).toEqual('function')
    expect(typeof passedProps.toggleTodoComplete).toEqual('function')
    expect(typeof passedProps.removeTodo).toEqual('function')

    passedProps.filterAll()
    passedProps.filterComplete()
    passedProps.filterIncomplete()
    passedProps.toggleTodoComplete('id_123')
    passedProps.removeTodo('id_123')

    expect(store.getActions()).toEqual([
      { type: 'test_set_filter', filter: 'all' },
      { type: 'test_set_filter', filter: 'complete' },
      { type: 'test_set_filter', filter: 'incomplete' },
      { type: 'test_toggle_complete', id: 'id_123' },
      { type: 'test_remove_todo', id: 'id_123' },
    ])

    expect(toJson(wrapper)).toMatchSnapshot()
  })
});
