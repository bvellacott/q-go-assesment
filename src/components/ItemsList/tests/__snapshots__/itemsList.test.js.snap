// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsList passes items and merged filter actions from store 1`] = `
<Component
  filterAll={[Function]}
  filterComplete={[Function]}
  filterIncomplete={[Function]}
  items={
    Array [
      1,
      2,
      3,
    ]
  }
  removeTodo={[Function]}
  setFilter={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  toggleTodoComplete={[Function]}
/>
`;

exports[`ItemsList renders filter button and correct handlers 1`] = `
<div>
  <ul
    className="filterList_actionBar"
  >
    <li>
      <FilterButton
        onClick={[Function]}
      >
        Show all
      </FilterButton>
    </li>
    <li>
      <FilterButton
        onClick={[Function]}
      >
        Show complete
      </FilterButton>
    </li>
    <li>
      <FilterButton
        onClick={[Function]}
      >
        Show incomplete
      </FilterButton>
    </li>
  </ul>
  <ul
    className="itemsList-ul"
  >
    <p
      id="items-missing"
    >
      Add some tasks above.
    </p>
  </ul>
</div>
`;
